### ISSUE TITLE
Add SortedDict.peekitem

### ISSUE THREAD

**Issue by bootandy** [grantjenks/sorted_containers#40]
We have popitem() but I'd like to have a peakitem() too.

So instead of having to do this:
key = mydict.iloc[0]
value = mydict[mydict.iloc[0]]

We can do this:
key,value = mydict.peakitem()

Happy to code it if you think it is suitable. If there is a nicer way to do the equivalent of peak() please let me know.

thanks.

**Comment by grantjenks** [comment]
Something like this?

```python
from sortedcontainers import SortedDict

class PeakableSortedDict(SortedDict):
    def peakitem(self):
        key = self.iloc[0]
        return key, self[key]

psd = PeakableSortedDict((val, val) for val in range(100))

psd.peakitem() # (0, 0)
```

There's little in the internals that could be leveraged to make this faster. Is there a reason you dislike inheriting and providing the method?

I've also had a desire for something like this. But I wanted the (key, value) pair at any index, like this:

```python
class BetterSortedDict(SortedDict):
    def item(self, index):
        key = self.iloc[index]
        return key, self[key]
```

Can you share a little more about your use case?

**Comment by bootandy** [comment]
Yes but I was thinking of adding peakitem directly in SortedDict instead of a subclass.

There is no reason I couldn't create a class that inherits from SortedDict. But when I was browsing the API and I saw there was a method called popitem I expected to see peakitem too. Then a colleague said exactly the same thing so I thought why not add it?

Personally when I see something that looks a bit like a queue and I see pop, I expect to see peak too.

Your example with item() also works. That would suit my use case well.

I don't know what your vision is for the library - so the above is just the opinions of a happy user.

**Comment by grantjenks** [comment]
I realized we both spelled "peek" as "peak". I think we mean "peek". Right?

As to vision, I want to make something useful but with a certain degree of minimalism. My rule of thumb is whether the current API can be used to efficiently solve the problem. This is kind of a border-case. Using internals, I could shave hundreds of nanoseconds off the peekitem implemented above. Is that really worth it? Depends on your use case. You're using SortedDict as a queue?

A little more background: I added popitem because it exists on collections.OrderedDict and blist.SortedDict. There's a good conversation on the python-ideas mailing list about OrderedDict supporting peekitem: https://mail.python.org/pipermail/python-ideas/2015-July/034487.html Most of the debate in the thread boils down to the general utility of the method and idioms: `next(iter(od))` and `next(reveresed(od))` which some think are more "pythonic".

I'm on the fence. Will others use SortedDict as a queue? The current `.iloc` often looks strange to newcomers who have never used Pandas `DataFrame.iloc`. Maybe an `item(index=-1)` method is better suited.

**Comment by bootandy** [comment]
Yes peek not peak. My bad sorry.

>I added popitem because it exists on collections.OrderedDict and blist.SortedDict. 
I thought so, that makes sense, it also explains why there is no peek.

I'd be happy with item(index=-1). I agree .iloc does look a little strange.

This isn't a big deal, as you point out it's very easy to implement a subclass that does the job. However I would expect other people to request either item() or peek() in the future. 

Feel free to close. Consider it food for thought anyway.

**Comment by grantjenks** [comment]
I'll leave it open with a "Feature" tag or something. It's great feedback. Thanks.

My other thought is to add an "IndexView" kind of like a "KeysView" or "ItemsView" that exposes numeric index lookup. Views are read-only so peekitem makes sense there.

To future readers: please "me too" or +1 if you're interested in the feature.

**Comment by grantjenks** [comment]
Implemented at cdba1b3. Will deploy at version 1.5.2 tomorrow.

**Comment by grantjenks** [comment]
Deployed at 1.5.2
